### PR TITLE
Add support for 2 point conical gradients

### PR DESCRIPTION
**Pull request by wmedrano**
- This is a straight port from Skia

Items remaining: None!
- ~Testing~
- ~Thinking about future of RadialGradient API or other simplifications~

### CODE REVIEW

**Comment by RazrFalcon** [comment]
Thanks! Looks good to me.

As for `RadialGradient` vs `ConicalGradient`, I think it's fine to break the API a bit by keeping the `RadialGradient` name and simply modifying the constructor. 

Keeping both at the same time would be too much work.

Other than that - if this change works fine for you and none of the [resvg](https://github.com/linebender/resvg) tests are failing after this change - we're ready to go.

**Comment by wmedrano** [comment]
Made the changes so that `RadialGradient` is now the 2 point conical gradient.

- I ran the resvg test suite
  - Using a focal radius of `0.0` produces no diffs.
    - https://github.com/wmedrano/resvg/commit/a6e9353fc0ce5f28737e590e74bbbb74769f2eae
    - The 3 diffs were pre-existing
  - Using the actual focal radius when available had 3 expected diffs
    - https://github.com/wmedrano/resvg/commit/80a8f7fff1e6fa10c9817ea7089ed382c4579072
    - The SVGs render as they do in Chrome/Chromium.
    - Firefox renders them slightly different.
- Suspicious change: https://github.com/linebender/tiny-skia/pull/164/commits/cef4b4362ff2fe242748c731453980420c6bda02
  - Required for the unit tests introduced to pass in avx.
  - https://github.com/linebender/tiny-skia/actions/runs/21232683569/job/61094164043
  - resvg test suite didn't require it.

**Comment by wmedrano** [comment]
Inspecting the 3 avx test failures its due to `NaN` related equality for data produced by XYTo2PtConicalStrip (new), XYTo2PtConicalSmaller (new), and XYTo2PtConicalGreater (pre-existing).

# conical_smaller_radial

```
stages = [SeedShader, Transform, XYTo2PtConicalSmaller, Mask2PtConicalDegenerates (diff), ...]
```

AVX got different output for `(3.1933205, 3.2251167, 3.2629573, 3.3092515, 3.3685663, 3.451985, 3.616074, NaN)` `!=` `(3.1933205, 3.2251167, 3.2629573, 3.3092515, 3.3685663, 3.451985, 3.616074, NaN)`
- normal result: `[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, NaN]`
- avx result: `[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]`

# strip_gradient

```
stages = [SeedShader, Transform, XYTo2PtConicalStrip, Mask2PtConicalNan (diff), ...]
```


AVX got different output for `(NaN, NaN, NaN, NaN, NaN, NaN, NaN, NaN)` != `(NaN, NaN, NaN, NaN, NaN, NaN, NaN, NaN)`
- normal result: `[NaN, NaN, NaN, NaN, NaN, NaN, NaN, NaN]`
- avx result: `[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]`

# conical_greater_radial

```
stages = [SeedShader, Transform, XYTo2PtConicalGreater, Mask2PtConicalDegenerates (diff), ...]
```

AVX got different output for `(NaN, NaN, NaN, NaN, NaN, NaN, NaN, NaN)` `!=` `(NaN, NaN, NaN, NaN, NaN, NaN, NaN, NaN)`
- normal result: `[NaN, NaN, NaN, NaN, NaN, NaN, NaN, NaN]`
- avx result: `[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]`

**Comment by wmedrano** [comment]
Any thoughts on how to move forward?

- Accept change to AVX instruction in `src/wide/f32x8_t.rs` to match non-avx
  - The unit tests then match Chrome/Firefox SVG rendering 
  - Seems to be what the pre-existing `mask_2pt_conical_degenerates` wanted. Which is to use `cmp_ne` to find `NaN`s.
- Use different unit tests and accept AVX may have differences in some scenarios
- Patch `XYTo2Pt.*` instructions to remove `NaNs`
  - NaN doesn't seem like that much of an edge case. It happens when `|y| > r0` in $$\sqrt{r_0^2 - y^2}$$ operation

**Comment by wmedrano** [comment]
This PR is good to go!

- The change to src/wide/f32x8_t.rs fixes an existing bug with `mask_2pt_conical_degenerates` that doesn't seem to be exercised by the existing unit tests or resvg.
  - Here it is reproduced in the main branch https://github.com/linebender/tiny-skia/compare/main...wmedrano:degen

**Comment by RazrFalcon** [comment]
Thanks! Looks good to me.
And thanks for the SIMD fix.

I would also add this to the changelog:
```
### Changed
- The `RadialGradient::new` requires a start radius now. Set the second argument to 0.0 to preserve the old behavior. 
```

Hopefully, the current maintainer would be able to merge and publish it soon.

PS:  there is one link to the original code left, not sure if intentionally or not.

File: src/pipeline/highp.rs [hunk 1]
```diff
@@ -1017,16 +1024,31 @@ fn xy_to_2pt_conical_greater(p: &mut Pipeline) {
...
+    let ctx = &p.ctx.two_point_conical_gradient;
+
+    let x = p.r;
+    let y = p.g;
+    p.r = x + (f32x8::splat(ctx.p0) - y * y).sqrt();
+
+    p.next_stage();
+}
+
+#[inline]
```
Comment (LaurenzV) [inline src/pipeline/highp.rs]: Is there a reason only this one is inlined and the other's arent?

Comment (wmedrano) [reply src/pipeline/highp.rs]: The non-inline functions are pipeline stages which need need to be real functions to be a part of the pipeline (https://github.com/linebender/tiny-skia/blob/9f5dd32f1dfa768ad7ac46427e4f740f23e7051f/src/pipeline/highp.rs#L59)

The inline ones are helpers. I moved it to the bottom with the other `#[inline]`. Also did some slight data shuffling since the helper is essentially [`cond_to_mask`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/src/opts/SkRasterPipeline_opts.h;l=2046;drc=bf712ec1a13783224debb691ba88ad5c15b93194) from big Skia.

**Comment by LaurenzV** [comment]
No need to move to them to the bottom, I just was curious. If I'm seeing correctly your newest commit reformatted a lot of unrelated code, can you revert that commit? Then I can merge.

**Comment by wmedrano** [comment]
Format changes reverted

**Review by LaurenzV — APPROVED** [review]
